accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,259 @@
+
+Using these data `pxe/terminal_entry` state would need to be adapted to look for pillar data in correct places and also to support generating single pxe entry or all pxe entries depending on the values passed to the state.
+
+## Generating PXE entries
+
+Update to routine is needed to support terminals connected directly to Uyuni server. Mainly calling `pxe/terminal_entry` state using salt runner or other means locally on Uyuni.
+
+Complete routine should look like:
+
+![Generating PXE entries flowchart](images/0000-retail-pxe-db-workflow.png)
+
+## Updating PXE entries
+
+Moving terminal between pxe servers happen. This means event handler needs to be aware of this possibility and adapt `pxeserver_id` field as needed.
+  
+Removal of the PXE server (branch server) system profile would set `pxeserver_id` to `null`. To update the entry terminal will need to be rebooted under new PXE server or user would need to manually update using API call.
+
+## API for manipulating and (re)generation of PXE entries
+
+Proposed API namespace `saltboot` as these entries related to saltbooted images.
+
+### API to modify PXE entries
+
+API should be implemented and it should be possible:
+
+* change pxeserver_id (use case when pxe server was changed - e.g. new machine registered as a replacement)
+* change kernel options for the terminals
+
+`saltboot.modifyPXEentries(sessionKey, [server_id1, server_id2, ...], { key: value})` where `key` is one of `pxeserver_id`, `kernel_parameters`, `boot_image`, `mac_address`
+
+### API to trigger regeneration of PXE entries
+
+API should be implemented and it should be possible to:
+
+* (re)generate all PXE entries for given pxe server (use case when pxe server was replaced, e.g. after hardware failure)
+* (re)generate single PXE entry for given pxe server and terminal (use case when changing the kernel option of the terminal, e.g. for debugging purposes)
+
+`saltboot.generatePXEentry(sessionKey, server_id)`
+
+`saltboot.generatePXEentries(sessionKey, pxeserver_id)`
+
+### API to show/list entries
+
+API is recommended to be implemented for sake of complete API based access and it should be possible to:
+
+* list PXE entries for given pxeserver id
+* show PXE entries for given server id
+
+`saltboot.listPXEentries(sessionKey, pxeserver_id)` -> list of `server_id` with pxe entry for given `pxeserver_id`
+
+`saltboot.getPXEentry(sessionKey, server_id)` -> dictionary of stored values for given `server_id`
+
+### API to remove PXE entries
+
+It is up to discussion as the entry should be automatically removed when terminal server is removed from Uyuni/SUSE Manager.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we **not** do this?
+
+  * obscure corner cases
+  
+    Currently unknown
+  
+  * will it impact performance?
+  
+    Multiple terminals booting at the same time - current state is handled solely by salt, proposed solution adds communication with database - increasing latency and some load on database
```

Comment:
How many is multiple worst case?

Reply:
I didn't do any measurement. I am just stating the obvious that java event handler and database communication enters the process of generating  pxe entries.

In case of salt-based system we are limited by salt resources (worker_threads), in this case we are also limited by (IIUC) tomcat and salt.